### PR TITLE
Improved dot printing and better cycle detection

### PR DESCRIPTION
**Pull request by Licenser**
## Description

This improves the dot printing and solves possible issues with an over-aggressive cycle detection (i.e. detecting cycles where there are none).

new

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

## Performance

| | main | this |
|-|-|-|
| real-workflow-throughput-json-v0.10 | 280-290 MB/s | 301-302 MB/s |

The cleanup in this seems to give a constant speedup of 3-6% with a significantly more dependable performance characteristic

image

### CODE REVIEW

File: rust-toolchain [hunk 1]
```diff
@@ -1 +1 @@
-1.51.0
+1.52.0
```
Comment (mfelsche) [inline rust-toolchain:1]: sneaky

**Review by mfelsche — APPROVED** [review]
i think there is 1 tiny clippy oversight, otherwise everything is good! :)

**Comment by Licenser** [comment]
There were >100 clippy oversights with some outdated allows, they get removed :D

**Review by mfelsche — DISMISSED** [review]

File: tremor-pipeline/src/query.rs [hunk 2]
```diff
@@ -479,6 +489,19 @@ impl Query {
                             .ok_or_else(|| Error::from("script not found"))?
                             .clone(),
                     ));
+
+                    let label = if let Stmt::ScriptDecl(s) = &inner_stmt {
+                        let e = s.extent(&query.node_meta);
+                        let mut h = Dumb::new();
+                        // We're trimming the code so no spaces are at the end then adding a newline
+                        // to ensure we're left justified (this is a dot thing, don't question it)
```
Comment (mfelsche) [inline tremor-pipeline/src/query.rs:497]: lol

**Review by mfelsche — APPROVED** [review]

**Comment by mfelsche** [comment]
Something seems to be fishy with an integration test not exiting: https://github.com/tremor-rs/tremor-runtime/tree/main/tremor-cli/tests/integration/query-exit

**Comment by Licenser** [comment]
I know, I dind't solve it today, fingers crossed for tomorrow.

File: tremor-pipeline/src/lib.rs [hunk 3]
```diff
@@ -197,12 +197,12 @@ pub(crate) fn common_cow(s: &str) -> beef::Cow<'static, str> {
...
 
 /// Type of nodes
-#[derive(Debug, Copy, Clone, Ord, PartialOrd, PartialEq, Eq, Hash)]
+#[derive(Debug, Clone, Ord, PartialOrd, PartialEq, Eq, Hash)]
 pub enum NodeKind {
     /// An input, this is the one end of the graph
     Input,
     /// An output, this is the other end of the graph
-    Output,
+    Output(Cow<'static, str>),
```
Comment (mfelsche) [inline tremor-pipeline/src/lib.rs:205]: I love that one! :)

Comment (Licenser) [reply tremor-pipeline/src/lib.rs:205]: Ja we were not tracking this correctly so it had to change :D

**Pull request merged**